build offliner schemas from json files

elfkuzco opened the pull request:
## Rationale
By dynamically building offliner schemas from json files, this PR lays the groundwork to define offliner flags and metadata
in data files, thus, allowing them to be defined in one place (TBD) and reduce redundancy of keeping definitions in sync
between scraper and zimfarm

## Changes
- build offliner schema from json files

elfkuzco commented:
A few comments regarding the json schema:
- `title` and `description` are required
- `min` and `max` should only be set for numeric types like `int`, `float`, etc
- `min_length` and `max_length` are for `strings`
- `pattern`s typically should be of Perl Compatible Regular Expressions as it could be consumed by different programming languages
- `choices` can either be a list of `{"title": ..., "value": ...}` or simply be  a list of strings. For cases where it is provided as a list of `{"title": ..., "value": ...}`, the `title` should be generally acceptable as an identifier across programming lanauges as it is used in constructing of the name of the enum member. If it is simply a list of strings, each string should satisfy the same constraint.
- The keys of the JSON schema should be generally acceptable as identifiers too as they constitute the parameters when constructing the `Offliner`. Provide an `alias` instead if a different name should be used.

elfkuzco commented on backend/src/zimfarm_backend/common/schemas/offliners/mwoffliner.json:
```diff
@@ -0,0 +1,268 @@
+    "type": "string",
+    "required": false,
+    "title": "Output folder",
+    "description": "Output folder for ZIM file or build folder. Leave it as `/output`",
```
Okay. This means, I would have to remove the `pattern` constraint on it too. Or is it safe to assume that other consumers of the JSON file (like the scrapers) can disregard the pattern?

elfkuzco commented:
yes, i think keeping it the files at the scraper level and using CI to release versions of the files.
given each scraper would probably have a version, would the zimfarm always be requesting the `latest` or `stable` version or will we have a way for pinning versions? for example, `mwoffliner` might be on `1.15` while `phet` is on `3.1`. or is this out of the question?

benoit74 commented:
> given each scraper would probably have a version, would the zimfarm always be requesting the latest or stable version or will we have a way for pinning versions? for example, mwoffliner might be on 1.15 while phet is on 3.1. or is this out of the question?

Would it be feasible and reasonable to have one pydantic model per scraper version, and select the appropriate one based on Docker image version of current recipe / task?

elfkuzco commented:
> > Would it be feasible and reasonable to have one pydantic model per scraper version, and select the appropriate one based on Docker image version of current recipe / task?
> 
> not really sure if that would work given the models are imported in order to define the `discriminator` field. Even if that were possible, I wonder how we would know which version to build. Unless we are going to set the version of each offliner as env vars and fetch that file version before starting. But that would only be one pydantic model.

benoit74 commented:
From discussions with kelson42:
- we are moving in the right direction
- the fact that scrapers are updating something in Zimfarm when CI runs seems to be correct
- we should probably rather store these definitions in DB (in JSON format)
- we need to have a system coping with multiple scraper version

More on this last point: it is now becoming clear that we must store in a recipe which scraper version this recipe is intended to use, and adapt flags (and their definitions) accordingly. We regularly wanna test `dev` version in the Zimfarm and are every now and then blocked by a missing flag / wrong definition (see https://github.com/openzim/zimfarm/issues/886). Not being to be warned of a problem when downgrading to previous worker version is also a problem.

@elfkuzco can you please have a second look at what it would take to be able to have a dynamic type definition **and** a dynamic list of allowed types? We are very open to any change needed to make this work.

I feel like if we create the `OfflinerSchema` type dynamically with `create_model` as well, it might work. Most probably static typing tools like pyright will not be able to validate anything ... but I don't mind much since on this field we precisely do not want static but dynamic typing.

elfkuzco commented:
I have created a table to store the offliner definitions. The table has the following columns:
- offliner
- definition
- version
- created_at

Also, I have created a foreign key from the schedule, task and requested task to this new table. This ensures that whenever any of the resource is finished, I know which version to retrieve and class schema to build at runtime. 

Right now, the PR should not be merged but can be reviewed. At least till I have the frontend set up. As it is, it fetches all offliner definitions on load. But this can no longer be done as we need to know which schedule to fetch first, retrieve it's definition before getting the definition.

Created migration too to seed the database with the initial version for all resources that are linked.

benoit74 commented on backend/src/zimfarm_backend/db/offliner_definition.py:
```diff
@@ -0,0 +1,161 @@
+
+def create_offliner_definition(
+    session: Session,
+    definition: OfflinerFlagSchema,
```
This `definition: OfflinerFlagSchema` argument makes me realize we probably have something wrong here in terms of naming.

I feel like we should find another name for both the field and the class.

What about switching from `definition` to `schema` (both in DB and in pydantic Schemas), and from `OfflinerFlagSchema` to `OfflinerSchema` (it can and will contain more than only flags at some point)

The `schema` is this dictionary. The `definition` is the whole object.

Or any other name, but reusing `definition` here makes things a bit confusing in the code.

benoit74 commented on backend/src/zimfarm_backend/migrations/versions/12096716adfd_add_offliner_definitions.py:
```diff
@@ -0,0 +1,167 @@
+        ["id"],
+    )
+
+    data_dir = Path(__file__).parent.parent / "seed"
```
Maybe rename folder from `seed` to `initial_offliner_definitions`

elfkuzco replied:
Done

benoit74 commented on backend/tests/test_validators.py:
```diff
@@ -745,72 +732,3 @@ def test_skipable_url_model_skip_validation(
         SecretUrlModel.model_validate(
             {"value": value}, context={"skip_validation": True}
         )
-
```
Where do we test this "slug validation" logic now?

benoit74 reviewed (commented):
Very well done, we are really on the good path. As you said, it must not be merged yet. I did not yet tested much the API behavior, but from the fact that you mostly left tests intact, it seems to be quite OK. I will test it a bit more once UI will be ready.

I do have some "little" concerns which needs to be addressed to be able to consider that work is really completed. I feel like it might be better to do it in later PRs, so that we can better document the decisions made.

I will only list concerns below for now. If needed, I do have ideas about how to address them (but again, it might be better to discuss this in a dedicated issue). Only first point is kinda important to solve soon, other point are a bit more "far-fetched".

- changes for nautilus and zimit relaxed schemas + specific validators on `xxx_schema_creator` (e.g. `lang_validator` for kolibri) + specific logic in `command_for`: I'm concerned about the fact that should flag name change, we will have to update Zimfarm code ; we need to find a solution to have this in offliner definition as well
- `offliner_id` field: now that we do not need it as a discriminator anymore, I feel like we moving it back to a "very top" level would help a lot ; but this is another breaking API change
- `xxx_schema_creator`s: I think we could move this into a specific new DB table as well, so that this is very extensible

elfkuzco commented on backend/src/zimfarm_backend/common/schemas/offliners/serializer.py:
```diff
@@ -217,7 +212,7 @@ def schema_to_flags(schema_class: type[BaseModel]) -> list[Flag]:
             description=description,
             key=field_alias,
-            label=title,
+            title=capitalize_words(title),
```
I used `title` at the application level so it closely mirrors the `FlagSchema` itself. The UI/clients expect `label` and I used a `serialization_alias` for that. As far as they are concerned, nothing changed.

Regarding the function, while editing the JSON files, I realized I forgot to make the `title` uniform on the UI. There are some places where words start with small letter and I figured this function would be nice to transform cases like this so things look nice.

elfkuzco commented on backend/src/zimfarm_backend/common/schemas/fields.py:
```diff
@@ -67,8 +58,10 @@ def not_empty(value: str) -> str:
 
 
-def validate_language_code(value: str, info: ValidationInfo) -> str:
+def validate_language_code(value: str | None, info: ValidationInfo) -> str | None:
```
Before we moved to this approach, the validation was enforced at the type level. For example, I had:
```py
ZIMLangCode = Annotated[
    str,
    Field(min_length=3, max_length=3),
    WrapValidator(skip_validation),
    AfterValidator(validate_language_code),
]


OptionalZIMLangCode = ZIMLangCode | None
```

The validator would never be called when the field is None because `None` has no validator. The validators were defined on the non-None types.

Now that we have moved to this approach, I only have some subset of types: `string`, `integer`, `float`, etc. And some of these fields are not required. So, this meant that the type would most likely be `string | None`. Thankfully, pydantic allows me to add some base constraints like `pattern`, `min_length` etc while creating the field for a particular flag. So, it won't call such inbuilt validations because those are None. Same thing with our concrete types.

Now, the issue is that we can't define functional/model validators in the JSON schema. And those would always be called by Pydantic. It's up to the author (me) to handle the case that these fields could be nullable and proceed accordingly. Essentially, I am having to do the same thing core types and pydantic `Field` do. Just more verbose.

elfkuzco commented:
> offliner_id field: now that we do not need it as a discriminator anymore, I feel like we moving it back to a "very top" level would help a lot ; but this is another breaking API change

I think it's used in a lot of more places too and it wasn't just serving as a discriminator. Maybe we could this in a later PR?

elfkuzco commented:
> Since we reuse this Offliner class everywhere, I think we should probably load all offliners at application startup (meaning we will have to reload application when we create a new offliner in DB ... this seems acceptable).

The `Offliner` class is an enum and while some of its helpers could be refactored as functions that take an argument, It is imported across functions too and used in places like https://github.com/openzim/zimfarm/blob/ef91be84720e4a60936a7482261859f1717b6803/backend/src/zimfarm_backend/utils/offliners.py#L29-L45

How do we handle things cases like this?
Also, you mention:
> For publisher, std_stats and std_output, we can simply add optional properties to mark flag to use:
If we allow the schema to have fields like 'is_publisher`, `is_std_stats`, etc.., I don't know how it could be used. What we are using the data for the fields is to build a pydantic `Field` and extra params like this are probably going to be excluded or raise a TypeError. 

One way I can think of is to extend the table to have columns like `std_stats`, `std_output`, `cmd` and use the data from there in the `command_for`

What do you think?

benoit74 commented:
`is_publisher`, `is_output` and `is_stats` must be "officially" defined in the `Flag` model.

`OFFLINER_DEFS` would be removed once we have have these properties directly on the flags. And this seems to be the correct way to do it because having it defined per offliner is wrong (like it was a simplification), this might change over the course of scraper changes. And having to modify Zimfarm when offliner became capable to write a stats file (for instance) is wrong, this information should come automatically with the offliner definition.

I already had a look at other places where we used this `Offliner` enum and I don't see any problematic hard to solve. When we adjust mount point for phet or docker config for zimit ; this would be way better in the offliner definition file (as high level config next to the flags) rather than directly in the code.

elfkuzco commented:
> is_publisher, is_output and is_stats must be "officially" defined in the Flag model

Okay. When I build up the field using `Field(is_publisher, is_output, is_stats, ....)`, this information is going to be lost. I can't do `offliner.config.<field>.is_publisher` because these values are not set by pydantic on the `Field`. 

Unless, I would have to use the definition while building the `command`.

elfkuzco commented:
- Added `stdOut` and `stdStats` to top level offliner definition to mirror the previous `OfflinerDefinition`
- Added `isPublisher` at the flag level. Used to determine which field is the publisher. Provided reasonable defaults
- Add Offliner model and remove corresponding enum
- Use map to prevent hitting the db to retrieve an offliner as it's used across a lot of places and it's likely going to be static data
- Created migration to load the DB with offliners

benoit74 reviewed (changes requested):
Thank you very much, I really like what you've achieve to build, this is definitely going to be a game changer. And to be honest, it really looks like it is working like a charm.

I feel like we could polish this PR a little bit more by:
- never returning `offliner_definition_id` in ORMs object (`TaskFullSchema`, `RequestedTaskFullSchema` and `ScheduleFullSchema`) but rather `offliner` and `version`, keeping the UUID an internal identifier
- this in turn would allow to get rid of the `GET /offliners/definition/{definition_id}` endpoint

WDYT?

What I see as currently broken / left to do before merging this PR (correct me if I miss something):
- for backward compatibility / transition, we need to maintain a `GET /offliners/{offliner_id}` endpoint: this will return the `initial` definition, allowing zimit-frontend to continue working.
- we cannot change offliner for a given schedule from UI, because it misses the combobox allowing to select version to use ; what about just "forcing" it to "initial" from the UI ; this could be sufficient until we've implemented the full UI
- confirm with two scrapers (1 Node.JS + 1 Python) that they can ingest the JSON file as well for their own purpose (CLI) ; probably mwoffliner for Node.JS ; do you wanna try this or is it better if I do this myself? scrapers are not that a dangerous beast after all ^^

What is left to do to close #886 (probably way better in distinct PRs, or even distinct issues and we close #886):
- implement the "version" combobox in "schedule edit" form
- show the "version" in schedule configuration + task configuration
- decide + implement what we wanna do when a scraper is released: endpoint to create new definition, permission to scope it to one offliner and only definition creation, what about upgrade/downgrade, is it defined in the schema as well and fully automated on release? or do we start with something manual as we currently do? could be linked to version changes from the user in the UI (same migrations are needed)
- decide + implement what we wanna show / edit in the UI in terms of offliner + definition (std_stats, full definition, ... nothing this is just technical data, ...)
- cleanup (drop "initial" defaults in some endpoints used for the transition)

elfkuzco commented:
> implement the "version" combobox in "schedule edit" form

created #1379 already for this.

elfkuzco commented:
> confirm with two scrapers (1 Node.JS + 1 Python) that they can ingest the JSON file as well for their own purpose (CLI) ; probably mwoffliner for Node.JS ; do you wanna try this or is it better if I do this myself? scrapers are not that a dangerous beast after all ^^

As per the meeting on Slack, it's simpler to ingest the JSON file in mwoffliner but not so much in python scrapers as those use argparse-specific parameters and behaviours like groups, dest, etc. One way to navigate around this without putting too much information in the JSON file would be to have specific logic at the scraper level to handle this. The zimfarm doesn't really do anything beyond building the arguments from existing flags and any additional metadata/information would not be particularly helpful in how it does it's job.

benoit74 commented:
Yes, I agree that we should probably not mind about the scrapers needing more logic to parse the JSON file for their own purpose. 

The fact that the JSON file will live inside scraper codebase is anyway already a big step forward, and scraper could even not parse it at all it will already help to sync changes in due time. 

Plus zimfarm anyway ignores extra fields, so scrapers are free to enrich the JSON file with "whatever needed" for their own purpose. We do not need to decide this at a central level.

Is this ready for a new review or do you still have open points to address?

elfkuzco commented:
- Added combobox to select version
- Added route to list all versions (used in populating combobox)
- Added handlers for when an offliner name changes and when it's version changes. Both cases lead to an update of the flags defintion.
- The watcher for the flags defintion rebuilds the flags by populating attributes that are set on the original schedule to the newer flags. This allows me to select an older version and still have some of the inputs across both versions to be used in populating the inputs as opposed to setting everything as empty. Same thing applies for offliners changed too although it almost always clears everything given offliners don't share fields.

elfkuzco commented:
> I finally wonder why we are asking for "Image name" in the UI since the backend is anyway forcing it to one specific value. Do I miss something?

The backend doesn't force it to a specific value. It only validates that the new image name is a valid image for the offliner. Plus, when the image name is changed, we update the tags for it. Do you want it to be removed?

benoit74 reviewed (approved):
LGTM

Thank you for this "very big" PR, glad we achieved to make this work.

There is some open point left but I will move this to dedicated issues as already agreed.
